### PR TITLE
CLIENTS-772: Multiple Contribution layouts

### PR DESCRIPTION
**Pull request by tomasstark**
This PR addresses [CLIENTS-772](https://laterpay.atlassian.net/browse/CLIENTS-772) and [CLIENTS-868](https://laterpay.atlassian.net/browse/CLIENTS-868).

- Introduces a new Contribution builder with collapsible preview of created Contribution.

- Adds a new [layout_type meta to Contribution offer](https://github.com/laterpay/revenue-generator/pull/272/files#diff-4f1ce13024716a0717ca35445447b2a592368487f03d00897332895737d3b300L82) with possible values of `box`, `button`, `bar`.

- Reorganizes templates for Contribution dialog a bit to make layout type readable from file name for easier navigation.

---

Most of the changes are related to templates, styles and JS part of the builder. The most juicy parts of this PR are:

- [Ditching dedicated 'Welcome' screen](https://github.com/laterpay/revenue-generator/pull/272/files#diff-d8de027c1cba3760a49ae26a8c803a096a58a26959c3ad3c9e77a6e4843aaf10L245) when adding a paywall for the first time since this logic is now moved directly to Edit Paywall screen in a form of modal.

- [Introducing a new constant](https://github.com/laterpay/revenue-generator/pull/272/files#diff-0beca18a469ce0ef11130f3a94b2f7f6997eecfa5f4f414442e163a0230e131cR82) to store a reference of plugin URL to easily point to that when adding new assets to front facing parts.

- [Adding a new post type](https://github.com/laterpay/revenue-generator/pull/272/files#diff-a198e823cba4e5e7ab0372e77ae8931ae98f891ee8e5f93713651d571dddeffb) for Contribution preview.
  - This acts like a helper post type to create a single post with [custom template](https://github.com/laterpay/revenue-generator/pull/272/files#diff-3d5a23010c63a18c1b9873a7a661555d01e17fcd3766ba5d2a1889454f594cd7) shown in a preview pane in contribution builder. Not exposed to public.

### CODE REVIEW

**Review by sfontenot — COMMENTED** [review]
Left some comments :grin:

File: assets/src/js/pages/revenue-generator-contribution.js [hunk 1]
```diff
@@ -9,1197 +9,973 @@
...
+
+				const validationType = $input.attr( 'data-validation' );
+				const isRequired = $input.attr( 'required' );
+				const value = $input.val();
+
+				switch ( validationType ) {
+					case 'url':
+						if ( value ) {
+							const test = value.match(
+								/(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g
```
[the remark this answers is not in the document]
Comment (tomasstark) [reply assets/src/js/pages/revenue-generator-contribution.js]: So this RegEx is something the plugin was using before for checking URL. I didn't come up with it so I can't exactly explain reasoning behind this decision. 🤷‍♂️  Now that you mention it, it's true the RegEx has its constraints.

However, it's worth mentioning that we don't need to check for perfect validity of URL here – it's just to prevent people typing in things like `this is not a url` or `321 not url` (so a random string) into the field. We don't need to check if the URL is a valid TLD or anything like that. As long as it *looks like URL*, we're happy. To check the validity of this URL (whether it's not a dead link etc.) is admin's responsibility while creating a contribution offer.

Since I don't like this ugly RegEx haunting our codebase though, I changed the code now to use [isURL module](https://github.com/laterpay/revenue-generator/blob/CLIENTS-772-multiple-contribution-layouts/assets/src/js/pages/revenue-generator-contribution.js#L209) in [@wordpress/url package](https://github.com/WordPress/gutenberg/tree/master/packages/url) that [we now use](https://github.com/laterpay/revenue-generator/blob/CLIENTS-772-multiple-contribution-layouts/package.json#L42). It's the official package that WordPress uses to check URLs in block editor and while not perfect either, it's at least aligned with the core principles now.

File: assets/src/js/pages/revenue-generator-contribution.js [hunk 2]
```diff
@@ -9,1197 +9,973 @@
...
 
-						// show loader.
-						showLoader();
+			setAmounts() {
+				const values = _( this.model.get( 'all_amounts' ) ).clone();
 
-						const allAmountJson = getContributionAmounts();
-						$o.allAmountsInput.val( allAmountJson );
+				for ( let i = 0; i < values.length; i++ ) {
+					values[ i ] = parseInt( values[ i ], 10 ) / 100;
```
[the remark this answers is not in the document]
Comment (tomasstark) [reply assets/src/js/pages/revenue-generator-contribution.js]: Yep, so this is a bit messy part that's harder to get at first. But essentially what we were doing in `setAmounts` method is parsing backend's contribution amounts for use in the builder UI. To explain a bit better, backend operates with `int` format (number of cents) and builder accepts floats so admins can enter `0.50` for 50 cents etc. so it's easier to work with that there.

However, I totally agree this double type casting is super ugly and I found that it's not even needed here. In [my recent commit](https://github.com/laterpay/revenue-generator/commit/2758e89b28ceed1ff51671bda3078131e62bc57b#diff-d8de027c1cba3760a49ae26a8c803a096a58a26959c3ad3c9e77a6e4843aaf10R320), I changed validation method to run the numbers through `parseFloat` only without compiling any object out of the values. All the other validation is moved to backend now, including adding information about type of revenue for that particular amount. When retrieving values for builder, we now have a WordPress filter in place that converts integer amounts stored in the database to floats accepted by builder.

To sum it up:

- We don't need to typecast to `int` and then to `float` in client side at all now.
- We validate all input in backend (much more reliable anyways).
- We have a filter in place that runs in the admin template to change amounts to builder readable format.

I think this is much cleaner approach. Let me know your thoughts! 😄

File: assets/src/js/pages/revenue-generator-contribution.js [hunk 3]
```diff
@@ -9,1197 +9,973 @@
...
-							hideLoader();
-						} );
+							const obj = {};
 
-						return false;
-					}
-				} );
+							if ( 0.0 < parseFloat( price ) ) {
+								price = price * 100;
+								obj.price = price;
```
[the remark this answers is not in the document]
Comment (tomasstark) [reply assets/src/js/pages/revenue-generator-contribution.js]: This is rendered obsolete now with https://github.com/laterpay/revenue-generator/commit/2758e89b28ceed1ff51671bda3078131e62bc57b#diff-d8de027c1cba3760a49ae26a8c803a096a58a26959c3ad3c9e77a6e4843aaf10 pushed.

But to elaborate on the comment, we wouldn't be able to use `0 < parseInt( price )` check here because user input can be `0.25` or `0.5` and `parseInt( 0.25 )` as well as `parseInt( 0.5 )` equals `0`. So the price wouldn't be added to the object while being completely valid price.

File: assets/src/js/pages/revenue-generator-contribution.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfontenot) [reply assets/src/js/pages/revenue-generator-contribution.js]: Ahhh, that makes sense, and sound good to me :grin:

File: assets/src/js/pages/revenue-generator-contribution.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sfontenot) [reply assets/src/js/pages/revenue-generator-contribution.js]: > I don't like this ugly RegEx haunting our codebase

:rofl:

**Review by sfontenot — APPROVED** [review]
lgtm
